Feature/post classes

FelixSFD opened the pull request:
## Introduced classes

### Post

Represents a post on StackExchange. Is the superclass of `Question` and `Answer`.

### Question/Answer

Both are inheriting from `Post`. The dictionary-initializers depend on `Post(dictionary:)`.


_Mentioning the rest would be too much. Just ask if anything is unclear._

## Changes in JsonHelper

Fixed bug that nested arrays and dictionaries were not parsed.

FelixSFD committed:
+ added basic implementations of Post and Question

FelixSFD committed:
+ added dictionary initializer to Post

FelixSFD committed:
+ Post.init(dictionary:) now converts some values to objects like URL and Date

FelixSFD committed:
+ added dictionary initializer for Question

FelixSFD committed:
+ added tests for Question
+ added JSON initializer for Post and Question

FelixSFD committed:
+ dictionary initializer of Question now saves the tags as [String]

FelixSFD committed:
+ added the Post.Notice struct

FelixSFD committed:
+ added the Question.ClosedDetails struct

FelixSFD committed:
+ added documentation for Question

FelixSFD committed:
+ added Comment class

FelixSFD committed:
Merge branch 'master' into feature/post-classes

FelixSFD committed:
+ made Post, Question and Comment use the classes and protocols introduced in pull-request #1

FelixSFD committed:
Merge branch 'master' into feature/post-classes

FelixSFD committed:
+ make use of JsonHelper and StringRepresentable
+ made ClosedDetails conform to JsonConvertible

FelixSFD committed:
- fixed Comment objects in dictionary representation of Post

FelixSFD committed:
- fixed missing properties in Question

FelixSFD committed:
+ added the Answer class and a basic test for it

FelixSFD committed:
+ added Question.answers

FelixSFD committed:
+ added the Site class (containing Site.Related, Site.Related.Relation, Site.State, Site.Type and Site.Styling)

FelixSFD committed:
+ added basic tests for the site class

FelixSFD committed:
- fixed typo

FelixSFD committed:
+ added Question.MigrationInfo

FelixSFD committed:
- fixed the Question.ClosedDetails.by_users property

jnkel commented on Sources/Answer.swift:
```diff
@@ -0,0 +1,90 @@
+        
+        self.can_flag = dictionary["can_flag"] as? Bool
+        
+        if let timestamp = dictionary["community_owned_cate"] as? Double {
```
Should that be `community_owned_date`?

jnkel replied:
Should that be `community_owned_date`?

jnkel commented on Sources/Answer.swift:
```diff
@@ -0,0 +1,90 @@
+            
+            for user in users {
+                let tmpUser = User(dictionary: user)
+                array.append(tmpUser)
```
We can get rid of the temporary variable here: `array.append(User(dictionary: user))`

FelixSFD replied:
yes, we could. But isn't it easier to read with the `let`?

jnkel replied:
I think it would be easier to read without it, but it doesn't really matter I guess.

jnkel commented on Sources/Comment.swift:
```diff
@@ -0,0 +1,124 @@
+ 
+ - seealso: [StackExchange API](https://api.stackexchange.com/docs/types/comment)
+ */
+public class Comment: JsonConvertible {
```
Can we make `Comment` a subclass of `Post`?  If not directly, can we make `Comment` and `Post` both subclasses of a parent class that contains the properties used by both?

FelixSFD replied:
`Post` has properties like `downvoted` which are not part of `Comment`.

Do you have an idea how to call a superclass of `Post` and `Comment`?

FelixSFD replied:
Maybe `Content`?

jnkel replied:
That's what I'm wondering.  I can't think of a good name.  I'll let you know if I think of one.

jnkel replied:
`Content` is a good idea

FelixSFD replied:
Okay. I'll try it

NobodyNada reviewed (changes requested):
Great job!  Just a couple of little things to consider before we merge this.

FelixSFD committed:
- fixed typo in community_owned_date

FelixSFD committed:
+ added Content as superclass of Post and Comment
- moved can_flag from Answer, Question and Comment to Content
- fixed missing Question.bounty_closes_date in Question.dictionary
- fixed missing Answer.question_id in Answer.init(dictionary:)
- moved Post.PostType outside of Post

FelixSFD committed:
+ added documentation comment for Content
